File: leet_code/733. Flood Fill.py
class Solution:
    def fill(self, image, rows, cols, r, c, cclr, nclr):
        if nclr == image[r][c]:
            return
        image[r][c] = nclr
        if r > 0 and image[r-1][c] == cclr:
            self.fill(image, rows, cols, r-1, c, cclr, nclr)
        if r < rows-1 and image[r+1][c] == cclr:
            self.fill(image, rows, cols, r+1, c, cclr, nclr)
        if c > 0 and image[r][c-1] == cclr:
            self.fill(image, rows, cols, r, c-1, cclr, nclr)
        if c < cols-1 and image[r][c+1] == cclr:
            self.fill(image, rows, cols, r, c+1, cclr, nclr)

    def floodFill(self, image, sr, sc, newColor):
        cclr = image[sr][sc]
        if cclr == newColor:
            return image
        self.fill(image, len(image), len(image[0]), sr, sc, cclr, newColor)
        return image


    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        
        rows = len(image)
        cols = len(image[0])
        clr = image[sr][sc]
        
        def dfs(y, x):
            if image[y][x] != clr or image[y][x] == newColor:
                return
        
            image[y][x] = newColor
            
            for ny, nx in [(y + 1, x), (y - 1, x), (y, x + 1), (y, x - 1)]:
                if not (0 <= ny < rows and 0 <= nx < cols):
                    continue
                
                dfs(ny, nx)
        
        dfs(sr, sc)
        return image


image = [[0,0,0],[0,1,1]]
sr = 1
sc = 1
newColor = 1

stime = time.time()
sol = Solution()
ret = sol.floodFill(image, sr, sc, newColor)
print('elapse time: {} sec'.format(time.time() - stime))
print(ret)

File: leet_code/733. Flood Fill.java
import java.util.Queue;
import java.util.Arrays;
import java.util.LinkedList;


class Solution {
    class Coord {
        public int y;
        public int x;
        
        Coord(int y, int x) {
            y = y;
            x = x;
        }
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int rows = image.length;
        int cols = image[0].length;
        int clr = image[sr][sc];
        
        Queue<Coord> q = new LinkedList<>();        
        q.add(new Coord(sr, sc));
        boolean [][] visited = new boolean[rows][cols];
        for(int i = 0; i < visited.length; i++) {
            Arrays.fill(visited[i], false);
        }
        
        int ycoords[] = {1, -1, 0, 0};
        int xcoords[] = {0, 0, 1, -1};
        
        while (!q.isEmpty()) {
            Coord res = q.poll();
            int y = res.y;
            int x = res.x;
            
            if ((image[y][x] != clr) || (image[y][x] == newColor) || (true == visited[y][x]))
                continue;
            
            image[y][x] = newColor;
            visited[y][x] = true;
            
            for (int i = 0; i < 4; i++) {
                int ny = y + ycoords[i];
                int nx = x + xcoords[i];
                
                if (!(0 <= ny && ny < rows && 0 <= nx && nx < cols))
                    continue;
            
                q.add(new Coord(ny, nx));
            }
        }
        
        return image;
    }

    void dfs(int[][] image, int y, int x, int clr, int newColor) {
        if (image[y][x] != clr || image[y][x] == newColor)
            return;
    
        image[y][x] = newColor;

        int ycoords[] = {1, -1, 0, 0};
        int xcoords[] = {0, 0, 1, -1};
        
        for (int i = 0; i < 4; i++) {
            int ny = y + ycoords[i];
            int nx = x + xcoords[i];
            if (!(0 <= ny && ny < rows && 0 <= nx && nx < cols))
                continue
            
            dfs(ny, nx);
        }
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int rows = image.length;
        int cols = image[0].length;
        int clr = image[sr][sc];

        
        dfs(image, sr, sc, clr, newColor);
    }

    public static void main(String[] args) {
    }
}